### PR TITLE
Update VecDeque CVE example

### PR DESCRIPTION
**Pull request by nchong-at-aws**
- Add README instructions for reproducing results
- Add a reproducible example that can be detected using valgrind
- Add a minimal example that Kani can error when using the buggy version of `VecDeque` and verify when using the fixed version
- Add a symbolic example that causes Kani to timeout
- Add an abstract implementation of `VecDeque` for deeper verification results

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: tests/cargo-kani/vecdeque-cve/Cargo.toml [hunk 1]
```diff
@@ -8,8 +8,3 @@ description = "Reproduces CVE-2018-1000657"
 
 [lib]
 path = "src/harness.rs"
-
-[dependencies]
-
-[package.metadata.kani.flags]
-cbmc-args = ["--no-built-in-assertions"]
```
Comment (nchong-at-aws) [inline tests/cargo-kani/vecdeque-cve/Cargo.toml:15]: I found this wasn't needed to verify the harnesses.

File: tests/cargo-kani/vecdeque-cve/src/harness.rs [hunk 2]
```diff
@@ -11,9 +11,16 @@
 #![feature(ptr_internals)]
 #![feature(rustc_allow_const_fn_unstable)]
 
+#[cfg(disable_debug_asserts)]
+macro_rules! debug_assert {
+    ( $( $x:expr ),* ) => {};
+}
```
Comment (nchong-at-aws) [inline tests/cargo-kani/vecdeque-cve/src/harness.rs:17]: I'm sure there's a better way using `RUSTFLAGS` but I couldn't work out a way to disable `debug_asserts` at the `cargo kani` command line.

Comment (zhassan-aws) [reply tests/cargo-kani/vecdeque-cve/src/harness.rs:17]: Kani overrides `debug_asserts` to apply the same instrumentation used for asserts: https://github.com/model-checking/kani/blob/4d4a5a7b99121f943b6d726b2c559767209531cd/library/std/src/lib.rs#L82 so what you did is likely the only way to disable them, unless we want to support disabling debug asserts (similar to the behavior of `cargo build --release`) which we can do through defining them as follows:
```
#[macro_export]
macro_rules! debug_assert {
    ($($x:tt)*) => (if $crate::cfg!(debug_assertions) ({ $crate::assert!($($x)*); })
}
```

Comment (nchong-at-aws) [reply tests/cargo-kani/vecdeque-cve/src/harness.rs:17]: Ah, I see. That's why my attempts didn't work. I tried `RUSTFLAGS='--cfg "debug_assertions=\"false\""'` and using a profile in `Cargo.toml` (https://doc.rust-lang.org/book/ch14-01-release-profiles.html):

```
[profile.dev]
opt-level = 3
```

File: tests/cargo-kani/vecdeque-cve/Cargo.toml (hunk 1 above)
Comment (nchong-at-aws) [reply tests/cargo-kani/vecdeque-cve/Cargo.toml:15]: @celinval Was there a reason why we removed these checks in the initial commit?

Comment (celinval) [reply tests/cargo-kani/vecdeque-cve/Cargo.toml:15]: I was trying to reduce the size of the report to make it more demo friendly. :)

File: tests/cargo-kani/vecdeque-cve/src/harness.rs (hunk 2 above)
Comment (zhassan-aws) [reply tests/cargo-kani/vecdeque-cve/src/harness.rs:17]: Would adding support for turning off `debug_assertions` help? I can push a PR to do that.

File: tests/cargo-kani/vecdeque-cve/Cargo.toml [hunk 3]
```diff
@@ -7,10 +7,4 @@ edition = "2018"
...
 
 [lib]
-path = "src/harness.rs"
-
-[dependencies]
-
-[package.metadata.kani.flags]
-enable-unstable = true
-cbmc-args = ["--no-built-in-assertions"]
+path = "src/harness.rs"
```
Comment (zhassan-aws) [inline tests/cargo-kani/vecdeque-cve/Cargo.toml:10]: Missing newline.

File: tests/cargo-kani/vecdeque-cve/src/abstract_vecdeque.rs [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+/// We don't model alloc type `A`
+struct AbstractVecDeque {
+    tail: usize,
+    head: usize,
+    buf: AbstractRawVec,
+}
+
+unsafe impl kani::Invariant for AbstractVecDeque {
+    fn is_valid(&self) -> bool {
+        self.tail < self.cap() && self.head < self.cap() && is_nonzero_pow2(self.cap())
```
Comment (zhassan-aws) [inline tests/cargo-kani/vecdeque-cve/src/abstract_vecdeque.rs:29]: Nice invariant!

File: tests/cargo-kani/vecdeque-cve/src/abstract_vecdeque.rs [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+fn handle_reserve(result: Result<(), TryReserveError>) {
+    match result.map_err(|e| e.kind()) {
+        Err(CapacityOverflow) => capacity_overflow(),
+        Err(AllocError) => handle_alloc_error(),
+        Ok(()) => { /* yay */ }
+    }
+}
+
+fn capacity_overflow() {
+    assert!(false);
```
Comment (zhassan-aws) [inline tests/cargo-kani/vecdeque-cve/src/abstract_vecdeque.rs:278]: Replace with `panic` or `unreachable`?

File: tests/cargo-kani/vecdeque-cve/src/abstract_vecdeque.rs [hunk 6]
```diff
@@ -0,0 +1,359 @@
...
+}
+
+fn handle_alloc_error() {
+    assert!(false);
```
Comment (zhassan-aws) [inline tests/cargo-kani/vecdeque-cve/src/abstract_vecdeque.rs:282]: Same for this one.

**Review by zhassan-aws — COMMENTED** [review]
Looks good!

**Pull request closed without merging**